neural_lam/datastore/mllam.py

```diff
@@ -0,0 +1,355 @@
+# Standard library
+import shutil
+import warnings
+from pathlib import Path
+from typing import List
+
+# Third-party
+import cartopy.crs as ccrs
+import mllam_data_prep as mdp
+import xarray as xr
+from loguru import logger
+from numpy import ndarray
+
+# Local
+from .base import BaseCartesianDatastore, CartesianGridShape
+
+
+class MLLAMDatastore(BaseCartesianDatastore):
+    """Datastore class for the MLLAM dataset."""
+
+    def __init__(self, config_path, n_boundary_points=30, reuse_existing=True):
+        """
+        Construct a new MLLAMDatastore from the configuration file at
+        `config_path`. A boundary mask is created with `n_boundary_points`
+        boundary points. If `reuse_existing` is True, the dataset is loaded
+        from a zarr file if it exists (unless the config has been modified
+        since the zarr was created), otherwise it is created from the
+        configuration file.
+
+        Parameters
+        ----------
+        config_path : str
+            The path to the configuration file, this will be fed to the
+            `mllam_data_prep.Config.from_yaml_file` method to then call
+            `mllam_data_prep.create_dataset` to create the dataset.
+        n_boundary_points : int
+            The number of boundary points to use in the boundary mask.
+        reuse_existing : bool
+            Whether to reuse an existing dataset zarr file if it exists and its
+            creation date is newer than the configuration file.
+
+        """
+        self._config_path = Path(config_path)
+        self._root_path = self._config_path.parent
+        self._config = mdp.Config.from_yaml_file(self._config_path)
+        fp_ds = self._root_path / self._config_path.name.replace(
+            ".yaml", ".zarr"
+        )
+
+        self._ds = None
+        if reuse_existing and fp_ds.exists():
+            # check that the zarr directory is newer than the config file
+            if fp_ds.stat().st_mtime > self._config_path.stat().st_mtime:
+                self._ds = xr.open_zarr(fp_ds, consolidated=True)
+            else:
+                logger.warning(
+                    "config file has been modified since zarr was created. "
+                    "recreating dataset."
+                )
+                shutil.rmtree(fp_ds)
```

Comment:
This is way too dangerous to do without any explicit user input. This could result in accidentally deleting terrabytes of data, that for some use cases would take a long time to recreate. I propose to just create a new zarr-archive, and leave it to the user to manually delete old, unneeded ones (could add text to the warning to explicitly instruct the user of this). A simple way to do this is to just rename the old one, perhaps based on its modification time.

Reply:
I have suggested on https://github.com/mllam/neural-lam/pull/66#discussion_r1730295168 to issue a warning based on the relative age of the config and dataset rather than the current delete+recreate implementation

Reply:
So then we don't recreate the dataset from an updated config? And if the user would want that (e.g. after seeing the warning) they would have to manually go in and delete/move the existing dataset first? I think that would be fine, just wanted to clarify.

Reply:
I have checked the code and removed the dangerous `shutil_rmtree`. The user will get a warning as suggested by @joeloskarsson. I will create a new issue to handle existing zarr archives in a smart way.
```
        if reuse_existing and fp_ds.exists():
            # check that the zarr directory is newer than the config file
            if fp_ds.stat().st_mtime > self._config_path.stat().st_mtime:
                self._ds = xr.open_zarr(fp_ds, consolidated=True)
            else:
                logger.warning(
                    "Config file has been modified since zarr was created. "
                    "The old zarr file will be used."
                    "To generate new zarr-file, please move the old one first."
                )
```